reader_concurrency_semaphore.cc

```diff
@@ -109,13 +111,20 @@ class reader_permit::impl
         if (_used_branches) {
             _state = reader_permit::state::active_used;
             on_permit_used();
+            if (_blocked_branches) {
+                _state = reader_permit::state::active_blocked;
+                on_permit_blocked();
```

Comment:
what does it mean for a permit to be simulteneaously marked both as used and blocked?

Reply:
Only used permits can be blocked. When a used permit becomes blocked, it is a signal that it no longer needs CPU to make progress: it needs something else, like I/O or a remote shard.
Unused permits don't participate in the CPU based admission.

Reply:
See https://github.com/scylladb/scylladb/blob/master/docs/dev/reader-concurrency-semaphore.md

Reply:
Makes sense. Thanks for the explanation

Reply:
Let's follow up with making the names clearer.

I suggest:
 - not_started
 - running
 - running_but_waiting_for_memory

Reply:
I do indeed plan on changing the names, I just wanted to wait for this PR to get in first, to avoid churn.